api/bucketsd/service.go

```diff
@@ -1497,6 +1518,367 @@ func (s *Service) PushPath(server pb.APIService_PushPathServer) (err error) {
+	doneCh chan<- addedFile,
+	errCh chan<- error,
+) (*fileAdder, error) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	fa, ok := q.q[pth]
+	if ok {
+		return fa, nil
+	}
+
+	key, err := addFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	reader, writer := io.Pipe()
+	fa = &fileAdder{
+		reader: reader,
+		writer: writer,
+	}
+	q.q[pth] = fa
+
+	eventCh := make(chan interface{})
+	chSize := make(chan string)
+	go func() {
+		for e := range eventCh {
+			event, ok := e.(*iface.AddEvent)
+			if !ok {
+				log.Error("unexpected event type")
+				continue
+			}
+			if event.Path != nil {
+				chSize <- event.Size // Save size for use in the final response
+			}
+		}
+	}()
+
+	var r io.Reader
+	if key != nil {
+		r, err = dcrypto.NewEncrypter(reader, key)
+		if err != nil {
+			return nil, fmt.Errorf("creating decrypter: %v", err)
+		}
+	} else {
+		r = reader
+	}
+
+	go func() {
+		defer close(eventCh)
+		res, err := ufs.Add(
+			ctx,
+			ipfsfiles.NewReaderFile(r),
+			options.Unixfs.CidVersion(1),
+			options.Unixfs.Pin(false),
+			options.Unixfs.Progress(true),
+			options.Unixfs.Events(eventCh),
+		)
+		if err != nil {
+			errCh <- fmt.Errorf("adding file: %v", err)
+			return
+		}
+		size := <-chSize
+		added, err := strconv.Atoi(size)
+		if err != nil {
+			errCh <- fmt.Errorf("getting file size: %v", err)
+			return
+		}
+		doneCh <- addedFile{path: pth, resolved: res, size: int64(added)}
+	}()
+
+	return fa, nil
+}
+
+func (s *Service) PushPaths(server pb.APIService_PushPathsServer) error {
+	log.Debugf("received push paths request")
+
+	ctx, cancel := context.WithCancel(server.Context())
+	defer cancel()
+
+	dbID, ok := common.ThreadIDFromContext(ctx)
+	if !ok {
+		return errDBRequired
+	}
+	dbToken, _ := thread.TokenFromContext(ctx)
+
+	req, err := server.Recv()
+	if err != nil {
+		return fmt.Errorf("on receive: %v", err)
+	}
+	var buckKey, buckRoot string
+	switch payload := req.Payload.(type) {
+	case *pb.PushPathsRequest_Header_:
+		buckKey = payload.Header.Key
+		buckRoot = payload.Header.Root
+	default:
+		return fmt.Errorf("push bucket path header is required")
+	}
+
+	lck := s.Semaphores.Get(buckLock(buckKey))
+	lck.Acquire()
+	defer lck.Release()
+
+	buck := &tdb.Bucket{}
+	err = s.Buckets.GetSafe(ctx, dbID, buckKey, buck, tdb.WithToken(dbToken))
+	if err != nil {
+		return fmt.Errorf("getting bucket: %v", err)
+	}
+	if buckRoot != "" && buckRoot != buck.Path {
+		return status.Error(codes.FailedPrecondition, buckets.ErrNonFastForward.Error())
+	}
+
+	var wg sync.WaitGroup
+	addedCh := make(chan addedFile)
+	doneCh := make(chan struct{})
+	errCh := make(chan error)
+	go func() {
+		queue := newFileQueue()
+		for {
+			req, err := server.Recv()
+			if err == io.EOF {
+				wg.Wait() // Request ended normally, wait for pending jobs
+				close(doneCh)
+				return
+			} else if err != nil {
+				errCh <- fmt.Errorf("on receive: %v", err)
+				return
+			}
+			switch payload := req.Payload.(type) {
+			case *pb.PushPathsRequest_Chunk_:
```

Comment:
Chunks from multiple files can be received at the same time. This enables the client code to determine the UX that makes sense for its use cases / environment. For example, a JS client may want to show a list of pending files that don't necessarily complete at the same time. In that case, progress could be also be shown on a per file basis. cc/ @carsonfarmer

Reply:
Cool, I think I have a nice idea for how to handle this that allows a caller to essentially pick and choose how they want to monitor progress. Nice.